File: typescript/hash/src/tree.ts
import { framedTuplePayload, hashRawBytes } from "./canonical.js";
import { hashError } from "./error.js";
import { concatBytes, hashCanonicalPayload, u64be } from "./hash.js";
import type {
  ArtifactTreeEntry,
  ArtifactTreeHashKind,
  HashId,
} from "./types.js";

const encoder = new TextEncoder();

export function validateArtifactPath(path: string): void {
  const invalid = (reason: string): never =>
    hashError("invalid-artifact-path", `invalid artifact path '${path}': ${reason}`);
  if (path.length === 0) invalid("path must contain at least one segment");
  if (path.startsWith("/") || path.endsWith("/")) {
    invalid("leading and trailing slashes are forbidden");
  }
  if (path.includes("//")) invalid("repeated slashes are forbidden");
  if (path.includes("\\")) invalid("backslashes are forbidden");
  if (path.includes("\0")) invalid("NUL bytes are forbidden");
  if (path.split("/").some((segment) => segment === "" || segment === "." || segment === "..")) {
    invalid("empty, '.' and '..' segments are forbidden");
  }
}

export function artifactTreePayload(entries: readonly ArtifactTreeEntry[]): Uint8Array {
  const sorted = [...entries];
  for (const entry of sorted) {
    validateArtifactPath(entry.path);
    if (entry.type === "symlink") {
      return hashError(
        "symlink-artifact",
        `artifact tree entries cannot be symlinks: '${entry.path}'`,
      );
    }
  }
  sorted.sort((left, right) => compareBytes(encoder.encode(left.path), encoder.encode(right.path)));

  const seen = new Set<string>();
  const payload: Uint8Array[] = [u64be(sorted.length)];
  for (const entry of sorted) {
    if (seen.has(entry.path)) {
      return hashError("duplicate-artifact-path", `duplicate artifact path '${entry.path}'`);
    }
    seen.add(entry.path);
    if (entry.type === "symlink") throw new Error("unreachable");
    const fileHash = hashRawBytes("artifact-file", entry.bytes);
    payload.push(
      framedTuplePayload([
        { label: "path", value: encoder.encode(entry.path) },
        { label: "fileHash", value: encoder.encode(fileHash) },
      ]),
    );
  }
  return concatBytes(...payload);
}

export function hashArtifactTree<K extends ArtifactTreeHashKind>(
  kind: K,
  entries: readonly ArtifactTreeEntry[],
): HashId<K> {
  return hashCanonicalPayload(kind, "artifact-tree-v1", artifactTreePayload(entries));
}

function compareBytes(left: Uint8Array, right: Uint8Array): number {
  const length = Math.min(left.length, right.length);
  for (let index = 0; index < length; index += 1) {
    const difference = (left[index] ?? 0) - (right[index] ?? 0);
    if (difference !== 0) return difference;
  }
  return left.length - right.length;
}

File: arete-hash/src/tree.rs
use std::collections::HashSet;

use crate::{
    framed_tuple_payload, hash_canonical_payload, hash_raw_bytes, require_profile, ArtifactFile,
    CanonicalizationProfile, HashError, HashId, Kind, TupleField,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ArtifactEntryKind {
    File,
    Symlink,
}

#[derive(Debug, Clone, Copy)]
pub struct ArtifactTreeEntry<'a> {
    pub path: &'a str,
    pub bytes: &'a [u8],
    pub kind: ArtifactEntryKind,
}

impl<'a> ArtifactTreeEntry<'a> {
    pub const fn file(path: &'a str, bytes: &'a [u8]) -> Self {
        Self {
            path,
            bytes,
            kind: ArtifactEntryKind::File,
        }
    }

    pub const fn symlink(path: &'a str) -> Self {
        Self {
            path,
            bytes: &[],
            kind: ArtifactEntryKind::Symlink,
        }
    }
}

pub fn validate_artifact_path(path: &str) -> Result<(), HashError> {
    let invalid = |reason| HashError::InvalidArtifactPath {
        path: path.to_string(),
        reason,
    };

    if path.is_empty() {
        return Err(invalid("path must contain at least one segment"));
    }
    if path.starts_with('/') || path.ends_with('/') {
        return Err(invalid("leading and trailing slashes are forbidden"));
    }
    if path.contains("//") {
        return Err(invalid("repeated slashes are forbidden"));
    }
    if path.contains('\\') {
        return Err(invalid("backslashes are forbidden"));
    }
    if path.contains('\0') {
        return Err(invalid("NUL bytes are forbidden"));
    }
    if path
        .split('/')
        .any(|segment| segment.is_empty() || segment == "." || segment == "..")
    {
        return Err(invalid("empty, '.' and '..' segments are forbidden"));
    }
    Ok(())
}

pub fn artifact_tree_payload(entries: &[ArtifactTreeEntry<'_>]) -> Result<Vec<u8>, HashError> {
    let mut entries = entries.to_vec();
    for entry in &entries {
        validate_artifact_path(entry.path)?;
        if entry.kind == ArtifactEntryKind::Symlink {
            return Err(HashError::SymlinkArtifact(entry.path.to_string()));
        }
    }
    entries.sort_by(|left, right| left.path.as_bytes().cmp(right.path.as_bytes()));

    let mut seen = HashSet::with_capacity(entries.len());
    let mut payload = Vec::new();
    payload.extend_from_slice(&(entries.len() as u64).to_be_bytes());
    for entry in entries {
        if !seen.insert(entry.path) {
            return Err(HashError::DuplicateArtifactPath(entry.path.to_string()));
        }
        let file_hash = hash_raw_bytes::<ArtifactFile>(entry.bytes)?.to_string();
        let leaf = framed_tuple_payload(&[
            TupleField::new("path", entry.path.as_bytes()),
            TupleField::new("fileHash", file_hash.as_bytes()),
        ])?;
        payload.extend_from_slice(&leaf);
    }
    Ok(payload)
}

pub fn hash_artifact_tree<K: Kind>(
    entries: &[ArtifactTreeEntry<'_>],
) -> Result<HashId<K>, HashError> {
    require_profile::<K>(CanonicalizationProfile::ArtifactTreeV1)?;
    let payload = artifact_tree_payload(entries)?;
    Ok(hash_canonical_payload(&payload))
}
